### ISSUE TITLE
Should omitempty check for Zero value when the field is a pointer?

### ISSUE THREAD

**Issue by robbrockbank** [go-playground/validator#249]
Hi - just wanted to get your take on the following behavior.

If a structure contains a pointer to (for example) an integer - it looks like the omitempty directive looks at both the pointer value (whether it is nil) and whether the pointed-to value is Zeroed.  I would *personally* have expected for a pointer that omitempty would just refer to whether the pointer was nil.

For example, suppose you have a structure with a pointer to an int - where the pointer may be nil (in which case you don't want it validated) - or it might point to a integer where you want to check it's range is, for example, 2-255.  I don't think this is possible at the moment without defining your own validator.

I would have expected a validation tag of "omitempty,gte=2,lte=255" to do the trick, but a value of 0 hits the omitempty directive and the gte tag is never used.

Confirmed with the following script:

```golang
package main

import (
	"fmt"
	"gopkg.in/go-playground/validator.v8"
)

type Test struct {
	MyInt *int `validate:"omitempty,gte=2,lte=255"`
}

func main() {
	val1 := 0
	val2 := 256
	t1 := Test{MyInt: &val1}   // This should fail validation on gte because value is 0
	t2 := Test{MyInt: &val2} // This should fail validate on lte because value is 256
	t3 := Test{MyInt: nil}  // This should succeed validation because pointer is nil

	config := &validator.Config{TagName: "validate"}
	validate := validator.New(config)
	err := validate.Struct(t1)
	fmt.Printf("t1: %v\n", err)
	err = validate.Struct(t2)
	fmt.Printf("t2: %v\n", err)
	err = validate.Struct(t3)
	fmt.Printf("t3: %v\n", err)
}
```

```
Roberts-MacBook-Pro-2:calico-go rob$ bin/temp
t1: <nil>
t2: Key: 'Test.MyInt' Error:Field validation for 'MyInt' failed on the 'lte' tag
t3: <nil>
```

**Comment by joeybloggs** [comment]
I will take a look at this asap

**Comment by joeybloggs** [comment]
Hey @robbrockbank 

ok I know why this is happening this way, let me explain:

here is from the docs:

### Omit Empty
Allows conditional validation, for example if a field is not set with a value (Determined by the "required" validator) then other validation such as min or max won't run, but if a value is set validation will run.
```
Usage: omitempty
```

so the key there is `determined by the "required" validator` and as I have mentioned before `required` will become `nondefault` in v9 someday; so the reason it's not failing on `gte` is because `required` is seeing `0` as the default value for the int **INSTEAD OF** `nil`. So coincidentally the only reason there is an issue is because the value is `0`, any other number would have proceeded as expected.

I totally agree, this is a **BUG** and should have behaved as you expected! I plan to keep track if the field in question's base type was a pointer, and if it was and we get to the lower `omitempty` logic then don't have to check `omitempty` because if we got there, there is already a value. 

I think this also may be an issue if MyInt was an `interface{}`

Thank You! Very Much! for reporting @robbrockbank this is the first legit bug found and to be fixed in validator in quite some time; I will have the FIX in before the end of the day.

**Comment by joeybloggs** [comment]
Hey @robbrockbank this should be fixed now via release v8.17.3

please update and let me know if this is fixed for you and we can get this issue closed! 😄

**Comment by robbrockbank** [comment]
Hi @joeybloggs!  Wow, that was a fast turnaround, thank you!  Yes I've confirmed with my example above that solves that problem.  Please feel free to close :-)

[probably not the right place to ask, but ... ] What's the timeline for updates to get into gopkg.in ?  I can pull from github, but all my imports in my code are pointing to gopkg.in ... of course perhaps the two are interchangeable in some golang magic that I'm not aware of (I'm a bit of a go novice I'm afraid).

**Comment by joeybloggs** [comment]
No problem @robbrockbank I not only maintain but actively use validator in a bunch of enterprise applications, so if there's a bug I try to squash it immediately for everyone, including myself ;)

ok gopkg.in, I'll try to answer as best I can.

So first off ''`go ge`" when hitting a gopkg.in URL is sent back what is essentially equivalent to , but not the same, as a redirect to an actual brach of your code.

So when you have more created any releases in your code, gopkg.in will grab the latest code in your repos default branch ( as I recall, you don't normally use it this way so...)

so you setup a branch in your code say `v1` then go to your releases tab in GitHub and create a release for branch v1. Now here's the most important part you have to use semantic versioning so your tag has to be in the form v1.0.0, or going.in will not recognise the release. And as soon as you do the gopkg.in knows to redirect to that specific branch with that release.

So in short to answer your question, gopkg.in will see the new released version as soon as you create it. Think of it as just redirecting `go get` to your released version. As for the godocs that gets refreshed at least once a day, but can force a manual refresh from the bottom of the godocs screen.

There is a bunch of documentation on Google.in here that may be more helpful http://labix.org/gopkg.in

Ya if you ever have any questions and don't want to create an issue, just contact me @ Dean.Karn@gmail.com

**Comment by jainparul9814** [comment]
Hi, this issue is still not fixed in v10
